pyccel/ast/internals.py

```diff
@@ -53,7 +53,7 @@ def is_elemental(self):
         return False
 
 
-class PyccelArraySize(PyccelInternalFunction):
+class NumpyArraySize(PyccelInternalFunction):
     """
     Class representing a call to a function which would
     return the shape of an object in a given dimension
```

Comment:
I am not convinced that `Numpy` needs to be in the name. In the pyccel/pyccel-cuda fork the same classes are used to handle the size of `CupyArray` objects. So perhaps the simplest solution is to remove the word `Numpy` from the class name and put both classes in ast/internals.py?

Reply:
If `CupyArray` objects also use the same classes, it makes a lot of sense to detach these from the Numpy extension!

Reply:
I am sorry but it appears that we have multiple inconsistencies, not only regarding names.

First of all, it appears that we have confused the names of the classes once again. The original `NumpyArraySize` (in the `devel` branch) returns the **total size** of an array and hence the class name **should not** be changed.

An exception to this behavior is the case where the `axis` argument is passed to the constructor, in which case an object of type `PyccelArraySize` (in the `devel` branch) is created. But I see no reason why `NumpyArraySize.__new__` should accept the `axis` argument in the first place, as we could simply create a `PyccelArraySize` object directly.

Finally, the name of `PyccelArraySize` **should** be changed to `NumpyArrayShapeElement` and not `NumpyArraySize`, because this class always represents the dimension of an array along a certain axis.

@mkaddani @EmilyBourne Do you agree with me?

------
As for the starting the class names with `Pyccel` instead of `Numpy` (or maybe removing this word altogether), this is a lesser problem which we can discuss later. Same goes for changing the location of these classes.

Reply:
I think that if we want to change the names then we should only do this once. It always takes a while to find all the places that need changing so it's easier to do while we already have a diff of those places.

The `axis` argument **must** appear in `NumpyArraySize.__new__` if we want the argument to be handled. Or new classes must be created. Built-in functions are handled by calling the constructor of a class whose arguments match the function. They are not called explicitly. If we really want to avoid this argument in this class then we would need a function similar to `pyccel.ast.numpyext.shape` which redirects the arguments to the correct class

Reply:
I agree that the names do not currently match the descriptions

Reply:
Just to clarify, the function `shape` of Numpy does **not** accept the `axis` argument (see [docs](https://numpy.org/doc/stable/reference/generated/numpy.shape.html#numpy.shape)).

Reply:
The shape **`property`** doesn't accept the `axis` argument but that is not the function that I get when I do `numpy.shape`. Instead it seems to be this one: https://numpy.org/devdocs/reference/generated/numpy.ma.size.html#numpy.ma.size
Which does have the `axis` argument. If this argument didn't exist then the tests would not run the python part

Reply:
Isn't that a function of the "masked arrays" module `numpy.ma`?

Reply:
I'm not really sure. Nothing else came up when I dug through the NumPy docs. But when I do:
```python
import numpy as np
np.shape?
```
The docs describe the `axis` argument

Reply:
That's very strange given that the tests run with those versions. Is there a difference between `np.shape` and from numpy import shape`?

Reply:
I will investigate in a virtual environment next week

Reply:
We seem to have gotten mixed up here. It is `size` which has the `axis` argument

Reply:
This is what comes of answering messages on my phone without checking things first :sweat_smile:

Reply:
The relevant numpy code is here:
https://github.com/numpy/numpy/blob/main/numpy/core/fromnumeric.py#L3218-L3252
I'm not sure how to get from there to the docs